### PR TITLE
When collapsing metrics via the collapse button in the menu bar, also…

### LINKED ISSUE #4790: When collapsing, collapse all items
When a user clicks the collapse button, in addition to all metrics, headers should also be collapsed.

Tasks:
- Change the popup ("Collapse all metrics and headers").
- Add the expansion state of headers to the URL query parameters so it can be controlled programmatically.
- Check that the collapse all button also works on the home page.

Out of scope
- Changing the icon, see #4395.

### PR DESCRIPTION
**Pull request by fniessink**
… collapse expanded headers. Closes #4790.

### CODE REVIEW

File: components/frontend/src/metric/MetricDetails.js [hunk 1]
```diff
@@ -83,6 +84,7 @@ export function MetricDetails({
     let any_error = last_measurement?.sources.some((source) => source.connection_error || source.parse_error);
     any_error = any_error || Object.values(metric.sources ?? {}).some((source) => !dataModel.metrics[metric.type].sources.includes(source.type))
     const sources_menu_item = any_error ? <Label color='red'>{"Sources"}</Label> : "Sources";
+    const metricUrl = `${window.location}#${metric_uuid}`
```
Comment (wkoot) [inline components/frontend/src/metric/MetricDetails.js:87]: Does this always work; is there never an anchor already specified?

File: components/frontend/src/header_footer/ResetSettingsButton.test.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    renderResetSettingsButton({
+        atReportsOverview: false,
+        handleDateChange: handleDateChange,
+        settings: settings
+    })
+    Object.values(settings).forEach((setting) => expect(setting.isDefault()))
+    fireEvent.click(screen.getAllByLabelText(/Reset this report's settings/)[0])
+    Object.values(settings).forEach((setting) => expect(setting.isDefault()))
+    expect(handleDateChange).not.toHaveBeenCalled()
+})
```
Comment (wkoot) [inline components/frontend/src/header_footer/ResetSettingsButton.test.js]:
```suggestion
})

```

File: components/frontend/src/header_footer/Menubar.js [hunk 3]
```diff
@@ -109,30 +112,16 @@ export function Menubar({
...
-                                        disabled={expandedItems.equals([])}
-                                        onClick={() => expandedItems.reset()}
-                                        icon={
-                                            <Icon
-                                                name={`caret ${expandedItems.equals([]) ? "right" : "down"}`}
-                                                size='large'
-                                            />
-                                        }
-                                        inverted
-                                        style={{ marginTop: -7, marginBottom: -7 }}  // Somehow the span removes the negative vertical margin the button has without the span, compensate
```
Comment (wkoot) [inline components/frontend/src/header_footer/Menubar.js:129]: Nice to get rid of all the additional workarounds

File: components/frontend/src/header_footer/CollapseButton.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import { Popup } from '../semantic_ui_react_wrappers';
+import { stringsURLSearchQueryPropType } from '../sharedPropTypes';
+
+export function CollapseButton({ expandedItems }) {
+    const label = "Collapse all headers and metrics"
+    return (
+        <Popup
+            on={["hover", "focus"]}
+            trigger={
+                <span  // We need a span here to prevent the popup from becoming disabled whenever the button is disabled
```
Comment (wkoot) [inline components/frontend/src/header_footer/CollapseButton.js:12]: Still needed? Is the empty line needed for comment or can it be `<span>`?

File: components/frontend/src/header_footer/CollapseButton.test.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        <CollapseButton
+            expandedItems={expandedItems ?? settings.expandedItems}
+        />
+    );
+}
+
+it("resets the expanded items", () => {
+    history.push("?expanded=tab")
+    const expandedItems = renderHook(() => useExpandedItemsSearchQuery())
+    expect(expandedItems.result.current.value).toStrictEqual(["tab"])
```
Comment (wkoot) [inline components/frontend/src/header_footer/CollapseButton.test.js:28]: Shouldn't we also test that opening the page with expanded=item does correctly display the item expanded?

File: components/frontend/src/header_footer/ResetSettingsButton.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import { optionalDatePropType, settingsPropType } from '../sharedPropTypes';
+
+export function ResetSettingsButton({ atReportsOverview, handleDateChange, reportDate, settings }) {
+    const metricsToHideDefault = atReportsOverview ? "all" : "none"
+    const label = `Reset ${atReportsOverview ? "reports overview" : "this report's"} settings`
+    return (
+        <Popup
+            on={["hover", "focus"]}
+            trigger={
+                <span  // We need a span here to prevent the popup from becoming disabled whenever the button is disabled
```
Comment (wkoot) [inline components/frontend/src/header_footer/ResetSettingsButton.js:13]: same

File: components/frontend/src/header_footer/ResetSettingsButton.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                            settings.showIssueUpdateDate.isDefault() &&
+                            settings.sortColumn.isDefault() &&
+                            settings.sortDirection.isDefault() &&
+                            settings.expandedItems.isDefault() &&
+                            reportDate === null
+                        }
+                        icon
+                        onClick={() => {
+                            handleDateChange(null);
+                            settings.dateInterval.reset();
```
Comment (wkoot) [inline components/frontend/src/header_footer/ResetSettingsButton.js]: Why not create a `settings.reset()` which does it all?

File: components/frontend/src/header_footer/CollapseButton.test.js (hunk 5 above)
Comment (wkoot) [reply components/frontend/src/header_footer/CollapseButton.test.js:28]: Or is this covered in `widgets/HeaderWithDetails.test.js`?

File: components/frontend/src/subject/SubjectTitle.js [hunk 8]
```diff
@@ -71,16 +73,23 @@ export function SubjectTitle({ atReportsOverview, report, subject, subject_uuid,
             render: () => <Tab.Pane><Share title="Subject permanent link" url={subjectUrl} /></Tab.Pane>
         }
     ];
+
     return (
         <HeaderWithDetails
             className="sticky"
             header={subjectTitle}
+            item_uuid={`${subject_uuid}:${tabIndex}`}
```
Comment (wkoot) [inline components/frontend/src/subject/SubjectTitle.js:81]: So we always include a tab index, which just defaults to the first one? The default http get query string is 1024 bytes, so does that mean we can expand about 20-25 tabs?

File: components/frontend/src/metric/MetricDetails.js (hunk 1 above)
Comment (fniessink) [reply components/frontend/src/metric/MetricDetails.js:87]: Ha, no, well-spotted!

New issue: https://github.com/ICTU/quality-time/issues/7761

File: components/frontend/src/header_footer/CollapseButton.js (hunk 4 above)
Comment (fniessink) [reply components/frontend/src/header_footer/CollapseButton.js:12]: `<span>  // Comment` doesn't work unfortunately.

File: components/frontend/src/header_footer/CollapseButton.test.js (hunk 5 above)
Comment (fniessink) [reply components/frontend/src/header_footer/CollapseButton.test.js:28]: Not really, I've added a test.

File: components/frontend/src/header_footer/ResetSettingsButton.js (hunk 7 above)
Comment (fniessink) [reply components/frontend/src/header_footer/ResetSettingsButton.js]: Yeah, I thought about it and then forgot. Will add it. And a `settings.areAllDefault()` method.

File: components/frontend/src/subject/SubjectTitle.js (hunk 8 above)
Comment (fniessink) [reply components/frontend/src/subject/SubjectTitle.js:81]: More. 49 expanded items was no problem: 

http://localhost:3000/example-report-quality-time?expanded=reports_overview%3A1&expanded_example-report=3ddbacd8-882c-4c0c-8f63-7f9fb00ade27%3A0,3edbace9-8e2c-4c1d-8e62-7a9fb40ae127%3A0,823001ba-d45f-4e91-83ba-5843708cb475%3A0,41b76e57-a3d8-4a65-9814-3df900f2f2ba%3A0,f7e1d0d7-37e1-42e4-8da1-5acdd48c65fc%3A0,a72882ff-36fb-40ea-a25e-7b92c5559317%3A0,848f2641-b170-4b55-8b7e-14f1ff283f1a%3A0,1b13d390-7521-48c4-a9ee-edfb2002d3a4%3A0,e7bfd848-869e-4e8e-b244-9e6568906973%3A0,fa330b0f-4775-4b5f-bf5b-aea9d13da3be%3A0,d551f09b-f89b-4b4d-a257-b9ae0bbbaf0d%3A0,de6f1067-0b64-4feb-a91c-e38b97146cb0%3A0,f854b9a4-715a-4083-b528-ce65ff7c229b%3A0,55d7177f-3d26-4f41-8bee-adc9fa7ffc2d%3A0,426631f9-3632-4c67-9769-ced54bdfac35%3A0,69d0dc8c-4eab-4198-bc8a-d2effe854b1f%3A0,3704a21e-b444-465d-a4a8-dcebcd8f487e%3A0,8865c5fc-6269-44b9-8924-095371fe537d%3A0,88151acf-0977-4ec0-8093-e9e8f41f409c%3A0,e3edc109-2f71-40cf-8825-0d9e8665bc0f%3A0,e3efc509-2a71-50df-8025-0aab8665bc0f%3A0,ee220264-9e66-49a4-981e-4dfe37a86a2c%3A0,example-report%3A0&expanded_example-report-next-action=6c7a5bc7-87c7-4ad3-9e60-824b9fda4ff9%3A0,26e24c34-0f79-4f82-bc9d-aeeac5c8a110%3A0,ff9f8448-9696-4184-82d0-caabdc0a8d49%3A0,00f83f0a-6551-469f-ae04-48f0b4f01760%3A0,428898d2-a514-4f43-992a-80211cf76707%3A0,84238917-7ed9-4340-91bb-8c2ca04b26df%3A0,e4248324-efa9-46a5-aa8e-910d6bbc6b0c%3A0,0dda60ef-f4f5-4cf6-87e7-fbf97256dbd0%3A0,1fbcbf0b-ee26-445f-b3eb-6d91ba0ee3ab%3A0,c8b3e6e7-4867-4b3d-aa12-6c0e2afedfe4%3A0,070c6abe-e16f-4360-8234-93cc58948017%3A0,517ef73e-6ba0-47fa-ab2b-345d12e63b85%3A0,2c453c91-9433-442d-857a-bb9d228f1070%3A0,example-report-next-action%3A0&expanded_example-report-quality-time=f51c0391-0a51-43f1-9bd1-69046d1acc06%3A0,b6fcc649-5873-48ac-8da3-61a18b3f055f%3A0,5a00e164-744c-4cd2-bf57-5dfca2c55a1f%3A0,22545de2-835f-4150-b9e8-d693e791429a%3A0,f7ac921d-0954-4701-858a-96845e637ed3%3A0,f95322e2-202c-4fee-93c2-46b4a6755fb6%3A0,25764450-478f-4225-8966-14ea79f8f7bd%3A0,62bbedcb-0197-4a10-97fa-690067c75120%3A0,1315f4e4-3acb-4412-abef-5904d76b4d35%3A0,ffb0f4a0-5b4a-44dc-9437-a3c2c2b4a8f9%3A0,79528152-36cc-4dfe-88db-18b021f0130f%3A0,79528152-36dc-48ff-89db-12b011ab1e04%3A0,8cd7fc6e-5a65-4958-83c3-8afcc461b628%3A0,example-report-quality-time%3A0

Comment (wkoot) [reply components/frontend/src/subject/SubjectTitle.js:81]: Ok, that's good to know. Did we not modify the max querystring size?

**Pull request merged**